python-package/lightgbm/sklearn.py

```diff
@@ -662,17 +670,57 @@ def __init__(
         self._n_classes: int = -1
         self.set_params(**kwargs)
 
+    # scikit-learn 1.6 introduced an __sklearn__tags() method intended to replace _more_tags().
+    # _more_tags() can be removed whenever lightgbm's minimum supported scikit-learn version
+    # is >=1.6.
+    # ref: https://github.com/microsoft/LightGBM/pull/6651
     def _more_tags(self) -> Dict[str, Any]:
         return {
             "allow_nan": True,
             "X_types": ["2darray", "sparse", "1dlabels"],
             "_xfail_checks": {
                 "check_no_attributes_set_in_init": "scikit-learn incorrectly asserts that private attributes "
                 "cannot be set in __init__: "
-                "(see https://github.com/microsoft/LightGBM/issues/2628)"
+                "(see https://github.com/microsoft/LightGBM/issues/2628)",
+                "check_n_features_in_after_fitting": (
+                    "validate_data() was first added in scikit-learn 1.6 and lightgbm"
+                    "supports much older versions than that"
```

Comment:
I think we should add in this comment that LightGBM supports [`predict_disable_shape_check=True`](https://lightgbm.readthedocs.io/en/latest/Parameters.html#predict_disable_shape_check) and we won't call `validate_data()` even after minimum sklearn version bump.

Reply:
hmmm I don't agree with that suggestion.

If `lightgbm`'s minimum floor on `scikit-learn` was `>=1.6`, I think we *could* safely consider calling `validate_data()`, maybe like:

```python
if not params.get("predict_disable_shape_check", True):
    self.validate_data(X)
```

LightGBM offering the ability to opt out of this validation wouldn't change that fact that the `check_n_features_in_after_fitting` test from `scikit-learn` would pass **by default**.

Reply:
Didn't think about such conditional approach. I like it!

Then, I think we can adopt `validate_data()` right now like we do with `_check_sample_weight()`:
https://github.com/microsoft/LightGBM/blob/59a3432b9d26290fcf25ba12b82feedd05384832/python-package/lightgbm/compat.py#L22-L30

We can try something like the following:
```python
try: 
    from sklearn.utils.validation import validate_data
except ImportError:
    # dummy function to support older version of scikit-learn
    def validate_data(_estimator, /, X='no_validation', y='no_validation', reset=True, validate_separately=False, skip_check_array=False, **check_params)
        check_array(X, y)
        check_X_y(X, y)
        _estimator._n_features_in = _estimator._n_features
        return X, y
```

Reply:
oh yeah, good idea!! I've started on this, still a bit of local testing to do... will try to push something here tomorrow.

Reply:
While testing this, I realized that LightGBM's `scikit-learn` interface **already** does not support the `predict_disable_shape_check` mechanism.

See this in `LGBMModel.predict()`.

https://github.com/microsoft/LightGBM/blob/d1d218c329b57e97b8938adfb1643e210ebe197c/python-package/lightgbm/sklearn.py#L1008-L1014

I think that behavior should be kept... if `scikit-learn`'s API does not support passing inputs to `predict()` with a different number of features than were present when the model was `fit()`, then neither should LightGBM's `scikit-learn` estimators.

Reply:
> I think that behavior should be kept... if `scikit-learn`'s API does not support passing inputs to `predict()` with a different number of features than were present when the model was `fit()`, then neither should LightGBM's `scikit-learn` estimators.

Agree!

Reply:
In recent commits, I've done the following:

* moved that number-of-features validation code out of `fit()` and into the `validate_data()` now defined in `compat.py`
* reworded the error message so it's identical to the one `sklearn.utils.validation.validate_data()` raises
* added a unit tests confirming that that error is raised, with that exact message, to be sure we'll find out if `scikit-learn` removes that check in the internals of `validate_data()` in the future